staking/slash/double-sign.go

```diff
@@ -1,52 +1,403 @@
+	)
+	errSlashDebtCannotBeNegative    = errors.New("slash debt cannot be negative")
+	errShardIDNotKnown              = errors.New("nil subcommittee for shardID")
+	errValidatorNotFoundDuringSlash = errors.New("validator not found")
+	zero                            = numeric.ZeroDec()
+	oneDoubleSignerRate             = numeric.MustNewDecFromStr("0.02")
+)
+
+// applySlashRate returns (amountPostSlash, amountOfReduction, amountOfReduction / 2)
+func applySlashRate(amount *big.Int, rate numeric.Dec) *big.Int {
+	return numeric.NewDecFromBigInt(
+		amount,
+	).Mul(rate).TruncateInt()
+}
+
+func payDownAsMuchAsCan(
+	snapshot, current *staking.ValidatorWrapper,
+	slashDebt, nowAmt *big.Int,
+	slashDiff *Application,
+) error {
+	if nowAmt.Cmp(common.Big0) == 1 && slashDebt.Cmp(common.Big0) == 1 {
+		// 0.50_amount > 0.06_debt => slash == 0.0, nowAmt == 0.44
+		if nowAmt.Cmp(slashDebt) >= 0 {
+			nowAmt.Sub(nowAmt, slashDebt)
+			if err := payDebt(
+				snapshot, current, slashDebt, slashDebt, slashDiff,
+			); err != nil {
+				return err
+			}
+		} else {
+			// 0.50_amount < 2.4_debt =>, slash == 1.9, nowAmt == 0.0
+			if err := payDebt(
+				snapshot, current, slashDebt, nowAmt, slashDiff,
+			); err != nil {
+				return err
+			}
+			nowAmt.Sub(nowAmt, nowAmt)
+		}
+	}
+
+	return nil
+}
+
+func delegatorSlashApply(
+	snapshot, current *staking.ValidatorWrapper,
+	rate numeric.Dec,
+	state *state.DB,
+	reporter common.Address,
+	doubleSignEpoch *big.Int,
+	slashTrack *Application,
+) error {
+
+	for _, delegationSnapshot := range snapshot.Delegations {
+		slashDebt := applySlashRate(delegationSnapshot.Amount, rate)
+		slashDiff := &Application{big.NewInt(0), big.NewInt(0)}
+		snapshotAddr := delegationSnapshot.DelegatorAddress
+		for _, delegationNow := range current.Delegations {
+			if nowAmt := delegationNow.Amount; delegationNow.DelegatorAddress == snapshotAddr {
+				l := utils.Logger().Info().RawJSON("delegation-snapshot", []byte(delegationSnapshot.String())).
+					RawJSON("delegation-current", []byte(delegationNow.String()))
+
+				l.Uint64("initial-slash-debt", slashDebt.Uint64()).
+					Str("rate", rate.String()).
+					Msg("attempt to apply slashing based on snapshot amount to current state")
+				// Current delegation has some money and slashdebt is still not paid off
+				// so contribute as much as can with current delegation amount
+				if err := payDownAsMuchAsCan(
+					snapshot, current, slashDebt, nowAmt, slashDiff,
+				); err != nil {
+					return err
+				}
+
+				// NOTE Assume did as much as could above, now check the undelegations
+				for _, undelegate := range delegationNow.Undelegations {
+					// the epoch matters, only those undelegation
+					// such that epoch>= doubleSignEpoch should be slashable
+					if undelegate.Epoch.Cmp(doubleSignEpoch) >= 0 {
+						if slashDebt.Cmp(common.Big0) <= 0 {
+							l.Msg("paid off the slash debt")
+							break
+						}
+						nowAmt := undelegate.Amount
```

Comment:
add a todo to remove the delegation that has 0 amount after paying debt.

Reply:
like doing an undelegate?